### ISSUE TITLE
dask DataFrame.read_parquet doesn't preserve the paths order

### ISSUE THREAD

**Issue by rajeee** [dask/dask#8829]
**What happened**:
Dask doesn't honor the order of files listed in the path

**What you expected to happen**:
I would have expected the partition ordering to be consistent with list of paths in read_parquet. Seems like the input path is automatically sorted internally without choice. 

**Minimal Complete Verifiable Example**:

```python
import dask.dataframe as dd
import numpy as np
df1=pd.DataFrame({'i': [1, 1], 'A': [1.0, 2.0], 'B': [11.0, 12.0]})
df1.to_parquet("df1.parquet")
df2=pd.DataFrame({'i': [0, 0], 'A':  [3.0, 4.0], 'B': [13.0, 14.0]})
df2.to_parquet("df2.parquet")
forward_df = dd.read_parquet(["df1.parquet", "df2.parquet"], engine='pyarrow')
reverse_df = dd.read_parquet(["df2.parquet", "df1.parquet"], engine='pyarrow')
forward_df.npartitions  # 2
reverse_df.npartitions  # 2

forward_df.get_partition(0).compute() # returns df1
forward_df.get_partition(0).compute() # Also returns df1. Expected to return df2


```

**Anything else we need to know?**:

**Environment**:

- Dask version:
- Python version:
- Operating System:
- Install method (conda, pip, source):

<details>
<summary>Cluster Dump State:</summary>

</details>

**Comment by rajeee** [comment]
I think this behavior is a limitation. If you want to load the files in a different order, there is currently no way to do so unless you rename the files. It makes sense to sort the paths if the list of paths is created by automatically listing the files. However, I think, it makes the most sense to not sort the user supplied list so that the user can load the files in the order they want. And it seems like it wouldn't be too hard to lift this limitation.

**Comment by bryanwweber** [comment]
> If you want to load the files in a custom order

Can you add something about why you want to do this? Perhaps there's another way.

**Comment by bryanwweber** [comment]
Can you use the `filters` keyword argument to filter the data, or the `columns` argument? It's not quite clear to me how your files are organized. Is it 1 house's timeseries per parquet file, or 1 house per column? How do you know which files are associated with each, say, city?

**Comment by rajeee** [comment]
It's one building per parquet file. The mapping of buildings to cities is stored in a separate file. I don't see how "filters" argument could help here.

**Comment by rajeee** [comment]
Thanks @jsignell.

I also had [another hurdle](https://stackoverflow.com/questions/71549122/how-to-read-list-of-parquets-with-partially-overlapping-set-of-columns-in-dask/71552863?noredirect=1#comment126500576_71552863) which required me to pre-process the files anyway. So I ended up renaming the files during the pre-process. 
But, yeah, for a general case, it would be nice if dask would honor the order of files in the input path list.

**Comment by rjzamora** [comment]
We can certainly avoid sorting the file names, but it couldn't be the default behavior (since current users/tests expect dask to correct input glob ordering to be numerical ordering).  Therefore, we would probably need to add another argument to `read_parquet`, and I get the sense that it is not an option many people would be using.  With that said, I will be happy to work on this if enough people can confirm that the current behavior is a problem.

If I understand correctly, the larger problem you are facing is that you need to aggregate many small files into a parquet dataset with fewer files. The most common approach to doing this for an arbitrary file pattern is definitely to use  `from_delayed`. However, in your SO question, I see that you are defining a distinct delayed function for every file. This should "work", but will result in a very large graph. If you already know the mapping between file names and “city”, for example, I’d want my delayed partitions to originate from a function designed to read in and concatenate data for multiple files (rather than just one).

If you want to try to use the `read_parquet` API for this, you can also try `read_parquet(..., aggregate_files=True)` (along with a `chunksize=` or `split_row_groups=` setting to specify the size of the output partitions).   However, this will not result in each partition cleanly mapping to the same city as you want (it will just give you fewer partitions).  I think the most common way to "partition" the parquet dataset on disk is to do something like:

`dd.read_parquet(paths, aggregate_files=True, chunksize=‘1GB’).to_parquet(partitioned_path, partition_on=["city"])`

This will not result in a smaller number of files, but you could always read the dataset back as

`dd.read_parquet(partitioned_path, aggregate_files="city")` to get a distinct city in each dask partition.

**Comment by rajeee** [comment]
Thanks for detailed reply @rjzamora. Yes, I also tried dataframe.from_delayed where each delayed object is returning a concatenation of a bunch of files. It did work okay but started breaking (taking forever to schedule) once the number of such delayed objects reached around 6K (happens when aggregating largish ~500K files). I finally transitioned to using a simple map of delayed objects that would directly write concatenated files to disk independently without forming a dataframe. 

I am personally not blocked by this anymore; so it's up to you guys to decide if adding an option to not sort the input path is a worthwhile feature to add or not. I am kinda wondering why the decision was made to sort the input path list in dask; if people want sorted path, it's not hard to sort in the application code before passing to dask.

**Comment by rjzamora** [comment]
>I finally transitioned to using a simple map of delayed objects that would directly write concatenated files to disk independently without forming a dataframe  

Yeah - If your only goal is to write a new dataset, then skipping the from_delayed part makes a lot of sense.  I know most people use delayed over a custom graph, but I personally do something like the following for simple cases like these:  

```python
import pandas as pd
from dask.dataframe.core import Scalar

files_per_task = 10
input_paths = ["tmpdir/part.0.parquet", "tmpdir/part.1.parquet", …]
output_dir = "new_tmpdir"

def read_concat_write(input_paths, output_path):
    pd.concat(
        [pd.read_parquet(path) for path in input_paths]
    ).to_parquet(output_path)

graph = {
    ("read-files", i): (
        read_concat_write,
        input_paths[j:j+files_per_task],
        f"{output_dir}/part.{i}.parquet",
    )
    for i, j in enumerate(range(0, len(input_paths), files_per_task))
}
graph[("finish", 0)] = (lambda x: None, list(graph.keys()))
Scalar(graph, "finish", "").compute()
```

>I am personally not blocked by this anymore; so it's up to you guys to decide if adding an option to not sort the input path is a worthwhile feature to add or not.

Regardless of it is currently blocking you, I am interested to know if the path-sorting has been a problem for anyone else.  This is the first issue I have seen about this, but it is certainly possible that others have run into the same problem but didn’t bother to share their experience on GitHub. (So, thank you for sharing - I encourage you to keep doing so when behavior seems “off”)

 >I am kinda wondering why the decision was made to sort the input path list in dask; if people want sorted path, it's not hard to sort in the application code before passing to dask.  

Fair question - The different engines have historically discovered paths in a slightly different order for directory-partitioned datasets. I think I remember this complicating aggregation behavior and parallel metadata parsing. The backends have also tended to use “glob” ordering instead of numerical ordering.  For these reasons, among others, the implicit sorting behavior was probably just introduced without much of an explicit decision. This is probably because no one ever asked for `read_parquet` to preserve the input path order, but people did ask for numerical ordering (and the directory-partitioning features that motivated the sorting).

All this is to say that: “Yes. It does seem to me like the user **should** have the option to preserve the order of the paths specified to read_parquet.  The reason that is isn’t supported is just that (1) no one has raised an issue about this yet, and (2) always sorting the paths has simplified other features in `read_parquet`.

**Comment by rajeee** [comment]
Thanks for the explanation and a bit of a history and context, Rick.

Your example code looks too raw for my experience (I don't have experience with raw task graph or Scalar construct), so I am using something like this:

```python
files_groups = np.array_split(input_paths, files_per_task)
output_paths = [f"{output_idr"/part.{i}.parquet" for i in range(len(files_groups))]
dask.compute(map(read_concat_write, files_groups, output_paths))
```
I assume this might be similar to your code though, under the hood.

I have created only a couple of issues so far in this repo, and the engagements and feedbacks have been awesome. This definitely increases my confidence in using dask on my data processing projects. So, kudos to you guys and keep up the good work!

**Comment by rjzamora** [comment]
Just a note here that #8852 will likely improve `from_delayed` performance for applications like this.
